File: app/services/exporter.py
"""SFT dataset export (BUILD_PLAN.md §7).

Three task shapes from one internal representation (§7.2), serialized to
three formats (§7.5), gated by the eligibility filters in §7.3, split
company-disjoint and stratified by operating_model (§7.4).

Constitution rule 4: these filters are correctness requirements, not
preferences - never relaxed to raise row count.
"""
import hashlib
from datetime import date, datetime
from typing import Any, Iterator, Optional

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.llm.prompts import REVIEWER_PROMPT_VERSION, REVIEWER_SYSTEM_PROMPT
from app.models import Company, HealthCheck, KillTrigger, Observation, ThesisVersion, TrainingSplit
from app.schemas.thesis import ThesisData

VALID_TASKS = {"thesis_synthesis", "verdict", "redline_extraction"}
VALID_FORMATS = {"anthropic", "openai", "llama"}
VALID_SPLITS = {"train", "eval", "all"}

_EVAL_HOLDOUT_BUCKET = 1500  # ~15% of 10000

SYSTEM_PROMPTS = {
    "thesis_synthesis": (
        "You are an investment analyst. Given raw company data (classification and "
        "prior operating metrics), produce a structured 7-pillar investment thesis. "
        "Return JSON only."
    ),
    "verdict": REVIEWER_SYSTEM_PROMPT,
    "redline_extraction": (
        "You are an investment analyst. Given an investment thesis narrative, extract "
        "the structured invalidation triggers (kill switches) it implies. Return JSON only."
    ),
}


def _to_float(value) -> Optional[float]:
    return float(value) if value is not None else None


def _hash_split(company_id: str) -> str:
    digest = int(hashlib.sha256(company_id.encode("utf-8")).hexdigest(), 16)
    return "eval" if (digest % 10000) < _EVAL_HOLDOUT_BUCKET else "train"


def ensure_split_assignments(db: Session) -> None:
    """Assigns any company missing a training_splits row. Deterministic on
    first assignment (hash of company_id); the row itself is then the
    source of truth so a later manual override survives future exports."""
    company_ids = set(db.scalars(select(Company.company_id)).all())
    existing = set(db.scalars(select(TrainingSplit.company_id)).all())
    for company_id in company_ids - existing:
        db.add(TrainingSplit(company_id=company_id, split=_hash_split(company_id)))
    db.commit()


def _company_ids_for_split(db: Session, split: str) -> set[str]:
    ensure_split_assignments(db)
    stmt = select(TrainingSplit.company_id)
    if split != "all":
        stmt = stmt.where(TrainingSplit.split == split)
    return set(db.scalars(stmt).all())


def _period_end_for(db: Session, company_id: str, period: str) -> Optional[date]:
    return db.scalar(
        select(Observation.period_end).where(
            Observation.company_id == company_id, Observation.period == period
        )
    )


def _thesis_synthesis_rows(db: Session, company_ids: set[str]) -> Iterator[dict[str, Any]]:
    if not company_ids:
        return
    versions = db.scalars(
        select(ThesisVersion).where(ThesisVersion.company_id.in_(company_ids))
    ).all()
    for version in versions:
        try:
            ThesisData.model_validate(version.thesis_data)
        except Exception:
            continue  # rule 3: must pass schema validation

        company = db.get(Company, version.company_id)
        # "Raw company data" input, honestly limited by the schema: classification
        # plus whatever was observed before this version was authored. This
        # platform has no filings/concall text store, so that's the real proxy
        # available - not a stand-in for it.
        prior_observations = db.scalars(
            select(Observation).where(
                Observation.company_id == version.company_id,
                Observation.period_end < version.authored_at,
            )
        ).all()
        yield {
            "task": "thesis_synthesis",
            "company_id": version.company_id,
            "input": {
                "company_name": company.name,
                "operating_model": company.operating_model,
                "currency": company.currency,
                "as_of": version.authored_at.date().isoformat(),
                "prior_observations": [
                    {"period": o.period, "metric_key": o.metric_key, "value": _to_float(o.numeric_value) or o.text_value}
                    for o in prior_observations
                ],
            },
            "output": version.thesis_data,
            "metadata": {"version_no": version.version_no, "authored_at": version.authored_at.isoformat()},
        }


def _verdict_rows(db: Session, company_ids: set[str], include_open: bool) -> Iterator[dict[str, Any]]:
    if not company_ids:
        return
    checks = db.scalars(
        select(HealthCheck).where(HealthCheck.company_id.in_(company_ids))
    ).all()
    for hc in checks:
        # rule 1: never train on unreviewed model output
        if hc.source == "ai_proposed" and not hc.human_confirmed:
            continue
        # rule 3: >= 3 reasoning steps
        if not hc.reasoning_chain or len(hc.reasoning_chain) < 3:
            continue

        company = db.get(Company, hc.company_id)
        # rule 4: default to resolved-outcome companies only for this task
        if not include_open and company.outcome == "open":
            continue

        version = db.get(ThesisVersion, hc.version_id)
        period_end = _period_end_for(db, hc.company_id, hc.period)
        # rule 2, the leakage check: the version must predate the period it's
        # reasoning about. No period_end on record means we can't verify it -
        # exclude rather than assume it's safe.
        if period_end is None or version.authored_at.date() >= period_end:
            continue

        triggers = db.scalars(
            select(KillTrigger).where(KillTrigger.version_id == hc.version_id)
        ).all()
        rule_engine_findings = [
            {
                "trigger": t.label,
                "metric_key": t.metric_key,
                "threshold": _to_float(t.threshold),
                "severity": t.severity,
            }
            for t in triggers
        ]

        yield {
            "task": "verdict",
            "company_id": hc.company_id,
            "input": {
                "thesis_data": version.thesis_data,
                "period": hc.period,
                "rule_engine_findings": rule_engine_findings,
            },
            "output": {
                "verdict": hc.verdict,
                "reasoning_chain": hc.reasoning_chain,
                "confidence": (hc.evidence or {}).get("confidence"),
            },
            "metadata": {
                "period": hc.period,
                "authored_at": version.authored_at.isoformat(),
                "period_end": period_end.isoformat(),
                "health_check_id": hc.id,
                "confidence": (hc.evidence or {}).get("confidence"),
            },
        }


def _redline_extraction_rows(db: Session, company_ids: set[str]) -> Iterator[dict[str, Any]]:
    if not company_ids:
        return
    versions = db.scalars(
        select(ThesisVersion).where(ThesisVersion.company_id.in_(company_ids))
    ).all()
    for version in versions:
        triggers = db.scalars(
            select(KillTrigger).where(KillTrigger.version_id == version.version_id)
        ).all()
        if not triggers:
            continue
        thesis_data = version.thesis_data or {}
        narrative = " ".join(
            [
                thesis_data.get("the_business", {}).get("what_it_does", ""),
                *thesis_data.get("the_growth_engine", []),
                thesis_data.get("the_big_change", {}).get("summary", ""),
                *thesis_data.get("why_we_believe_it", []),
            ]
        ).strip()
        if not narrative:
            continue

        yield {
            "task": "redline_extraction",
            "company_id": version.company_id,
            "input": {"thesis_narrative": narrative},
            "output": [
                {
                    "label": t.label,
                    "metric_key": t.metric_key,
                    "operator": t.operator,
                    "threshold": _to_float(t.threshold),
                    "severity": t.severity,
                    "action": t.action,
                    "grace_periods": t.grace_periods,
                    "manual_check": t.manual_check,
                }
                for t in triggers
            ],
            "metadata": {"version_no": version.version_no, "authored_at": version.authored_at.isoformat()},
        }


def _rows_for_task(db: Session, task: str, company_ids: set[str], include_open: bool) -> list[dict[str, Any]]:
    if task == "thesis_synthesis":
        return list(_thesis_synthesis_rows(db, company_ids))
    if task == "verdict":
        return list(_verdict_rows(db, company_ids, include_open))
    if task == "redline_extraction":
        return list(_redline_extraction_rows(db, company_ids))
    raise ValueError(f"unknown task {task!r}")


def _serialize(row: dict[str, Any], fmt: str) -> dict[str, Any]:
    system_prompt = SYSTEM_PROMPTS[row["task"]]
    user_content = row["input"] if isinstance(row["input"], str) else row["input"]
    metadata = {
        **row["metadata"],
        "task": row["task"],
        "company_id": row["company_id"],
        "prompt_version": REVIEWER_PROMPT_VERSION,
    }

    if fmt == "anthropic":
        return {
            "system": system_prompt,
            "messages": [
                {"role": "user", "content": user_content},
                {"role": "assistant", "content": row["output"]},
            ],
            "metadata": metadata,
        }
    if fmt == "openai":
        return {
            "messages": [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": user_content},
                {"role": "assistant", "content": row["output"]},
            ],
            "metadata": metadata,
        }
    if fmt == "llama":
        return {
            "prompt": f"<s>[INST] <<SYS>>\n{system_prompt}\n<</SYS>>\n\n{user_content} [/INST]",
            "completion": row["output"],
            "metadata": metadata,
        }
    raise ValueError(f"unknown format {fmt!r}")


def export_rows(
    db: Session, task: str, fmt: str, split: str = "all", include_open: bool = False
) -> list[dict[str, Any]]:
    if task not in VALID_TASKS:
        raise ValueError(f"task must be one of {sorted(VALID_TASKS)}")
    if fmt not in VALID_FORMATS:
        raise ValueError(f"format must be one of {sorted(VALID_FORMATS)}")
    if split not in VALID_SPLITS:
        raise ValueError(f"split must be one of {sorted(VALID_SPLITS)}")

    company_ids = _company_ids_for_split(db, split)
    rows = _rows_for_task(db, task, company_ids, include_open)
    return [_serialize(r, fmt) for r in rows]


def eval_split_verdict_rows(db: Session, include_open: bool = False) -> list[dict[str, Any]]:
    """Pre-serialization 'verdict' task rows for the eval-split companies -
    the input the eval harness (eval/run_eval.py) actually needs, without
    reaching into this module's underscore-prefixed internals."""
    company_ids = _company_ids_for_split(db, "eval")
    return list(_verdict_rows(db, company_ids, include_open))


def export_stats(db: Session, split: str = "all", include_open: bool = False) -> dict[str, Any]:
    ensure_split_assignments(db)
    company_ids = _company_ids_for_split(db, split)

    row_counts: dict[str, int] = {}
    class_balance: dict[str, dict[str, int]] = {}
    by_operating_model: dict[str, dict[str, int]] = {}
    leakage_violations = 0

    companies_by_id = {c.company_id: c for c in db.scalars(select(Company)).all()}

    for task in VALID_TASKS:
        rows = _rows_for_task(db, task, company_ids, include_open)
        row_counts[task] = len(rows)

        if task == "verdict":
            balance: dict[str, int] = {}
            for r in rows:
                v = r["output"]["verdict"]
                balance[v] = balance.get(v, 0) + 1
            class_balance[task] = balance

        model_counts: dict[str, int] = {}
        for r in rows:
            company = companies_by_id.get(r["company_id"])
            key = company.operating_model if company else "unknown"
            model_counts[key] = model_counts.get(key, 0) + 1
        by_operating_model[task] = model_counts

    # Leakage check across ALL verdict-eligible rows regardless of the include_open
    # filter, since this is a correctness audit, not a dataset-shape query.
    for r in _verdict_rows(db, _company_ids_for_split(db, "all"), include_open=True):
        authored_at = datetime.fromisoformat(r["metadata"]["authored_at"]).date()
        period_end = date.fromisoformat(r["metadata"]["period_end"])
        if authored_at >= period_end:
            leakage_violations += 1

    split_counts = {
        s: len(_company_ids_for_split(db, s)) for s in ("train", "eval")
    }

    return {
        "row_count": sum(row_counts.values()),
        "row_count_by_task": row_counts,
        "class_balance": class_balance,
        "by_operating_model": by_operating_model,
        "leakage_violations": leakage_violations,
        "companies_by_split": split_counts,
    }

File: web/src/lib/server/services/exporter.ts
// Ports app/services/exporter.py. Three task shapes from one internal
// representation, serialized to three formats, gated by eligibility filters,
// split company-disjoint by a deterministic hash of company_id.
//
// These filters are correctness requirements, not preferences - never relax
// them to raise row count.
import { and, eq, inArray, lt } from 'drizzle-orm';
import { createHash } from 'node:crypto';
import { db } from '../db';
import { companies, healthChecks, killTriggers, observations, thesisScenarios, thesisVersions, trainingSplits } from '../db/schema';
import { thesisData as thesisDataSchema } from '../schemas/thesis';
import { REVIEWER_PROMPT_VERSION, REVIEWER_SYSTEM_PROMPT } from '../llm/prompts';

export const VALID_TASKS = new Set(['thesis_synthesis', 'verdict', 'redline_extraction']);
export const VALID_FORMATS = new Set(['anthropic', 'openai', 'llama']);
export const VALID_SPLITS = new Set(['train', 'eval', 'all']);

const EVAL_HOLDOUT_BUCKET = 1500; // ~15% of 10000

const SYSTEM_PROMPTS: Record<string, string> = {
	thesis_synthesis:
		'You are an investment analyst. Given raw company data (classification and ' +
		'prior operating metrics), produce a structured 7-pillar investment thesis. ' +
		'Return JSON only.',
	verdict: REVIEWER_SYSTEM_PROMPT,
	redline_extraction:
		'You are an investment analyst. Given an investment thesis narrative, extract ' +
		'the structured invalidation triggers (kill switches) it implies. Return JSON only.'
};

function toFloat(value: unknown): number | null {
	return value != null ? Number(value) : null;
}

/** sha256(company_id) % 10000 < 1500 => eval, else train. Deterministic. */
export function hashSplit(companyId: string): 'train' | 'eval' {
	const digest = createHash('sha256').update(companyId, 'utf-8').digest('hex');
	const mod = BigInt('0x' + digest) % 10000n;
	return mod < BigInt(EVAL_HOLDOUT_BUCKET) ? 'eval' : 'train';
}

/** Assigns any company missing a training_splits row. Deterministic on first
 * assignment (hash of company_id); the row itself is then the source of
 * truth so a later manual override survives future exports. */
export async function ensureSplitAssignments(): Promise<void> {
	const companyRows = await db.select({ companyId: companies.companyId }).from(companies);
	const existingRows = await db.select({ companyId: trainingSplits.companyId }).from(trainingSplits);
	const existing = new Set(existingRows.map((r) => r.companyId));
	const missing = companyRows.map((r) => r.companyId).filter((id) => !existing.has(id));
	for (const companyId of missing) {
		await db.insert(trainingSplits).values({ companyId, split: hashSplit(companyId) });
	}
}

async function companyIdsForSplit(split: string): Promise<Set<string>> {
	await ensureSplitAssignments();
	const rows =
		split === 'all'
			? await db.select({ companyId: trainingSplits.companyId }).from(trainingSplits)
			: await db
					.select({ companyId: trainingSplits.companyId })
					.from(trainingSplits)
					.where(eq(trainingSplits.split, split));
	return new Set(rows.map((r) => r.companyId));
}

async function periodEndFor(companyId: string, period: string): Promise<string | null> {
	const [row] = await db
		.select({ periodEnd: observations.periodEnd })
		.from(observations)
		.where(and(eq(observations.companyId, companyId), eq(observations.period, period)))
		.limit(1);
	return row?.periodEnd ?? null;
}

export interface ExportRow {
	task: string;
	company_id: string;
	input: unknown;
	output: unknown;
	metadata: Record<string, unknown>;
}

async function thesisSynthesisRows(companyIds: Set<string>): Promise<ExportRow[]> {
	if (!companyIds.size) return [];
	const versions = await db.select().from(thesisVersions).where(inArray(thesisVersions.companyId, [...companyIds]));
	const rows: ExportRow[] = [];
	for (const version of versions) {
		const parsed = thesisDataSchema.safeParse(version.thesisData);
		if (!parsed.success) continue; // rule 3: must pass schema validation

		const [company] = await db.select().from(companies).where(eq(companies.companyId, version.companyId)).limit(1);
		// "Raw company data" input, honestly limited by the schema: classification
		// plus whatever was observed before this version was authored. No
		// filings/concall text store exists, so this is the real proxy available.
		const priorObservations = await db
			.select()
			.from(observations)
			.where(and(eq(observations.companyId, version.companyId), lt(observations.periodEnd, toDateOnly(version.authoredAt))));

		rows.push({
			task: 'thesis_synthesis',
			company_id: version.companyId,
			input: {
				company_name: company.name,
				operating_model: company.operatingModel,
				currency: company.currency,
				as_of: toDateOnly(version.authoredAt),
				prior_observations: priorObservations.map((o) => ({
					period: o.period,
					metric_key: o.metricKey,
					value: toFloat(o.numericValue) ?? o.textValue
				}))
			},
			output: version.thesisData,
			metadata: { version_no: version.versionNo, authored_at: version.authoredAt.toISOString() }
		});
	}
	return rows;
}

function toDateOnly(d: Date): string {
	return d.toISOString().slice(0, 10);
}

async function verdictRows(companyIds: Set<string>, includeOpen: boolean): Promise<ExportRow[]> {
	if (!companyIds.size) return [];
	const checks = await db.select().from(healthChecks).where(inArray(healthChecks.companyId, [...companyIds]));
	const rows: ExportRow[] = [];
	for (const hc of checks) {
		// rule 1: never train on unreviewed model output
		if (hc.source === 'ai_proposed' && !hc.humanConfirmed) continue;
		// rule 3: >= 3 reasoning steps
		const reasoningChain = hc.reasoningChain as unknown[] | null;
		if (!reasoningChain || reasoningChain.length < 3) continue;

		// rule 4: default to resolved-outcome scenarios only for this task -
		// "outcome" is per-user thesis state, not company-wide.
		const [scenario] = await db.select().from(thesisScenarios).where(eq(thesisScenarios.id, hc.scenarioId)).limit(1);
		if (!includeOpen && scenario.outcome === 'open') continue;

		const [version] = await db.select().from(thesisVersions).where(eq(thesisVersions.versionId, hc.versionId)).limit(1);
		const periodEnd = await periodEndFor(hc.companyId, hc.period);
		// rule 2, the leakage check: the version must predate the period it's
		// reasoning about. No period_end on record means we can't verify it -
		// exclude rather than assume it's safe.
		if (periodEnd == null || toDateOnly(version.authoredAt) >= periodEnd) continue;

		const triggers = await db.select().from(killTriggers).where(eq(killTriggers.versionId, hc.versionId));
		const ruleEngineFindings = triggers.map((t) => ({
			trigger: t.label,
			metric_key: t.metricKey,
			threshold: toFloat(t.threshold),
			severity: t.severity
		}));

		const evidence = (hc.evidence ?? {}) as { confidence?: number | null };
		rows.push({
			task: 'verdict',
			company_id: hc.companyId,
			input: {
				thesis_data: version.thesisData,
				period: hc.period,
				rule_engine_findings: ruleEngineFindings
			},
			output: {
				verdict: hc.verdict,
				reasoning_chain: reasoningChain,
				confidence: evidence.confidence ?? null
			},
			metadata: {
				period: hc.period,
				authored_at: version.authoredAt.toISOString(),
				period_end: periodEnd,
				health_check_id: hc.id,
				confidence: evidence.confidence ?? null
			}
		});
	}
	return rows;
}

interface ThesisDataShape {
	the_business?: { what_it_does?: string };
	the_growth_engine?: string[];
	the_big_change?: { summary?: string };
	why_we_believe_it?: string[];
}

async function redlineExtractionRows(companyIds: Set<string>): Promise<ExportRow[]> {
	if (!companyIds.size) return [];
	const versions = await db.select().from(thesisVersions).where(inArray(thesisVersions.companyId, [...companyIds]));
	const rows: ExportRow[] = [];
	for (const version of versions) {
		const triggers = await db.select().from(killTriggers).where(eq(killTriggers.versionId, version.versionId));
		if (!triggers.length) continue;
		const thesisData = (version.thesisData ?? {}) as ThesisDataShape;
		const narrative = [
			thesisData.the_business?.what_it_does ?? '',
			...(thesisData.the_growth_engine ?? []),
			thesisData.the_big_change?.summary ?? '',
			...(thesisData.why_we_believe_it ?? [])
		]
			.join(' ')
			.trim();
		if (!narrative) continue;

		rows.push({
			task: 'redline_extraction',
			company_id: version.companyId,
			input: { thesis_narrative: narrative },
			output: triggers.map((t) => ({
				label: t.label,
				metric_key: t.metricKey,
				operator: t.operator,
				threshold: toFloat(t.threshold),
				severity: t.severity,
				action: t.action,
				grace_periods: t.gracePeriods,
				manual_check: t.manualCheck
			})),
			metadata: { version_no: version.versionNo, authored_at: version.authoredAt.toISOString() }
		});
	}
	return rows;
}

async function rowsForTask(task: string, companyIds: Set<string>, includeOpen: boolean): Promise<ExportRow[]> {
	if (task === 'thesis_synthesis') return thesisSynthesisRows(companyIds);
	if (task === 'verdict') return verdictRows(companyIds, includeOpen);
	if (task === 'redline_extraction') return redlineExtractionRows(companyIds);
	throw new Error(`unknown task ${task}`);
}

export function serialize(row: ExportRow, fmt: string): Record<string, unknown> {
	const systemPrompt = SYSTEM_PROMPTS[row.task];
	const userContent = row.input;
	const metadata = {
		...row.metadata,
		task: row.task,
		company_id: row.company_id,
		prompt_version: REVIEWER_PROMPT_VERSION
	};

	if (fmt === 'anthropic') {
		return {
			system: systemPrompt,
			messages: [
				{ role: 'user', content: userContent },
				{ role: 'assistant', content: row.output }
			],
			metadata
		};
	}
	if (fmt === 'openai') {
		return {
			messages: [
				{ role: 'system', content: systemPrompt },
				{ role: 'user', content: userContent },
				{ role: 'assistant', content: row.output }
			],
			metadata
		};
	}
	if (fmt === 'llama') {
		return {
			prompt: `<s>[INST] <<SYS>>\n${systemPrompt}\n<</SYS>>\n\n${userContent} [/INST]`,
			completion: row.output,
			metadata
		};
	}
	throw new Error(`unknown format ${fmt}`);
}

export async function exportRows(
	task: string,
	fmt: string,
	split: string = 'all',
	includeOpen = false
): Promise<Record<string, unknown>[]> {
	if (!VALID_TASKS.has(task)) throw new Error(`task must be one of ${[...VALID_TASKS].sort().join(', ')}`);
	if (!VALID_FORMATS.has(fmt)) throw new Error(`format must be one of ${[...VALID_FORMATS].sort().join(', ')}`);
	if (!VALID_SPLITS.has(split)) throw new Error(`split must be one of ${[...VALID_SPLITS].sort().join(', ')}`);

	const companyIds = await companyIdsForSplit(split);
	const rows = await rowsForTask(task, companyIds, includeOpen);
	return rows.map((r) => serialize(r, fmt));
}

/** Pre-serialization 'verdict' task rows for the eval-split companies - the
 * input an eval harness would need, without reaching into this module's
 * internal row builders. */
export async function evalSplitVerdictRows(includeOpen = false): Promise<ExportRow[]> {
	const companyIds = await companyIdsForSplit('eval');
	return verdictRows(companyIds, includeOpen);
}

export interface ExportStats {
	row_count: number;
	row_count_by_task: Record<string, number>;
	class_balance: Record<string, Record<string, number>>;
	by_operating_model: Record<string, Record<string, number>>;
	leakage_violations: number;
	companies_by_split: Record<string, number>;
}

export async function exportStats(split: string = 'all', includeOpen = false): Promise<ExportStats> {
	await ensureSplitAssignments();
	const companyIds = await companyIdsForSplit(split);

	const rowCounts: Record<string, number> = {};
	const classBalance: Record<string, Record<string, number>> = {};
	const byOperatingModel: Record<string, Record<string, number>> = {};
	let leakageViolations = 0;

	const companyRows = await db.select().from(companies);
	const companiesById = new Map(companyRows.map((c) => [c.companyId, c]));

	for (const task of VALID_TASKS) {
		const rows = await rowsForTask(task, companyIds, includeOpen);
		rowCounts[task] = rows.length;

		if (task === 'verdict') {
			const balance: Record<string, number> = {};
			for (const r of rows) {
				const v = (r.output as { verdict: string }).verdict;
				balance[v] = (balance[v] ?? 0) + 1;
			}
			classBalance[task] = balance;
		}

		const modelCounts: Record<string, number> = {};
		for (const r of rows) {
			const company = companiesById.get(r.company_id);
			const key = company ? company.operatingModel : 'unknown';
			modelCounts[key] = (modelCounts[key] ?? 0) + 1;
		}
		byOperatingModel[task] = modelCounts;
	}

	// Leakage check across ALL verdict-eligible rows regardless of the
	// include_open filter, since this is a correctness audit, not a
	// dataset-shape query.
	const allCompanyIds = await companyIdsForSplit('all');
	for (const r of await verdictRows(allCompanyIds, true)) {
		const authoredAt = String(r.metadata.authored_at).slice(0, 10);
		const periodEnd = String(r.metadata.period_end);
		if (authoredAt >= periodEnd) leakageViolations++;
	}

	const splitCounts: Record<string, number> = {};
	for (const s of ['train', 'eval']) {
		splitCounts[s] = (await companyIdsForSplit(s)).size;
	}

	return {
		row_count: Object.values(rowCounts).reduce((a, b) => a + b, 0),
		row_count_by_task: rowCounts,
		class_balance: classBalance,
		by_operating_model: byOperatingModel,
		leakage_violations: leakageViolations,
		companies_by_split: splitCounts
	};
}
